### PR TITLE
feat(cli) Add new CLI

### PR DESCRIPTION
**Pull request by etnoy**
This PR adds a brand new CLI to Immich!

Features supported:
* Get server information
* Upload files
* Import files
* Delete files after upload
* Hash files before upload

Missing features so far (for future PRs)
* Add assets to album

Other consequences:
We should rename the "admin" cli to something less confusing now that both CLI are in the same repo
Until this new CLI has become more stable we should keep current docs for old cli

### CODE REVIEW

File: cli/package.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    "@typescript-eslint/eslint-plugin": "^5.60.1",
+    "chai": "^4.3.7",
+    "commander": "^11.0.0",
+    "eslint": "^8.43.0",
+    "eslint-config-prettier": "^8.8.0",
+    "eslint-plugin-jest": "^27.2.2",
+    "eslint-plugin-prettier": "^4.2.1",
+    "eslint-plugin-unicorn": "^47.0.0",
+    "glob": "^10.3.1",
+    "immich-sdk": "^1.62.0",
```
Comment (alextran1502) [inline cli/package.json:26]: Now that the CLI is in the same repo, can we make the openAPI generator generates the SDK directly in the repo, similar to the web? This would keep the API in sync with the server changes.

Comment (etnoy) [reply cli/package.json:26]: Yes, we should do that, but I'm not sure where to start. Do you want to give me a pointer or two?

Comment (etnoy) [reply cli/package.json:26]: Will do!

Comment (etnoy) [reply cli/package.json]: Done

File: cli/src/services/upload.service.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        'Content-Type': 'application/json',
+      },
+    };
+  }
+
+  public checkIfAssetAlreadyExists(path: string, checksum: string): Promise<any> {
+    this.checkAssetExistenceConfig.data = JSON.stringify({ assets: [{ id: path, checksum: checksum }] });
+
+    // TODO: retry on 500 errors?
+    return axios(this.checkAssetExistenceConfig);
```
Comment (alextran1502) [inline cli/src/services/upload.service.ts:49]: can we use the SDK for this?

Comment (etnoy) [reply cli/src/services/upload.service.ts:49]: The upload endpoint just didn't work, I tried for hours as you might recall. The bulk upload check though, that might work

File: server/bin/generate-open-api.sh [hunk 3]
```diff
@@ -23,11 +23,23 @@ function web {
...
+  cd ../../..
+  npx --yes @openapitools/openapi-generator-cli generate -g typescript-axios -i ./immich-openapi-specs.json -o ../cli/src/api/open-api -t ./openapi-generator/templates/cli --additional-properties=useSingleRequestParameter=true
+}
+
 if [[ $1 == 'mobile' ]]; then
   mobile
 elif [[ $1 == 'web' ]]; then
   web
+elif [[ $1 == 'cli' ]]; then
+  web
```
Comment (alextran1502) [inline server/bin/generate-open-api.sh:40]: Do you mean to have this as `cli` instead of `web`

File: cli/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
+A command-line interface for interfacing with Immich
```
Comment (alextran1502) [inline cli/README.md:1]: Can you please add some notes for how to use this in development? ts-node is not a dev-dependency, correct? You would have to install it separately

**Comment by alextran1502** [comment]
After testing, I have some thoughts, let me know how you want to address them

1. We should have the yes/no prompt before actually performing the action for user consent.
2. We should have the statistic of how many files were uploaded and how many files will be uploaded so that the user can keep track of the directory.
3. Printing out each line of assets might be redundant if you are uploading 100,000 assets. It would be better to write the skipped file and error file into a log file.

The issue with file extension is addressed on Discord.

File: server/bin/generate-open-api.sh (hunk 3 above)
Comment (etnoy) [reply server/bin/generate-open-api.sh:40]: Duh, of course

File: cli/src/services/upload.service.ts (hunk 2 above)
Comment (etnoy) [reply cli/src/services/upload.service.ts:49]: I've spent some time but not getting the SDK to work properly at this time. I suggest we keep the current implementation and focus on the SDK in the future

File: cli/README.md (hunk 4 above)
Comment (etnoy) [reply cli/README.md:1]: I must admit I'm not knowledgeable enough with ts to know the best way to run a script. Perhaps you can advise?

**Comment by etnoy** [comment]
> After testing, I have some thoughts, let me know how you want to address them
> 
> 1. We should have the yes/no prompt before actually performing the action for user consent.

There's a --dry-run option to take care of this case without having to mess with interactive responses mid-way

> 2. We should have the statistic of how many files were uploaded and how many files will be uploaded so that the user can keep track of the directory.

Done.
> 3. Printing out each line of assets might be redundant if you are uploading 100,000 assets. It would be better to write the skipped file and error file into a log file.

Done. Added a progress bar.

File: cli/src/services/upload.service.ts (hunk 2 above)
Comment (alextran1502) [reply cli/src/services/upload.service.ts:49]: I meant for the check if asset is exist endpoint. We can use axios for the upload endpoint

File: cli/README.md (hunk 4 above)
Comment (jrasm91) [reply cli/README.md:1]: `npx ts-node cli/src` would probably work. I think,  though, eventually all these notes should get moved to the documentation side, right?

File: cli/jest.config.ts
Comment (jrasm91) [inline cli/jest.config.ts:1]: In general, some configuration files can be inlined in the package.json file directly. We might want to revisit that in the future. Generally, I prefer that when possible as it reduces the number of dot files needed at the project root.

File: cli/package.json [hunk 5]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "immich-cli",
+  "dependencies": {
+    "axios": "^1.4.0",
+    "form-data": "^4.0.0",
+    "mime-types": "^2.1.35",
+    "systeminformation": "^5.18.4"
```
Comment (jrasm91) [inline cli/package.json:7]: I don't think we need this dependency.

File: cli/package.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    "picomatch": "^2.3.1",
+    "shx": "^0.3.4",
+    "ts-jest": "^29.1.0",
+    "ts-node": "^10.9.1",
+    "tslib": "^2.5.3",
+    "typescript": "^4.9.4",
+    "yaml": "^2.3.1"
+  },
+  "scripts": {
+    "build": "shx rm -rf dist && tsc -b",
```
Comment (jrasm91) [inline cli/package.json:42]: What is `shx`?

File: cli/src/commands/server-info.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  static enableJsonFlag = true;
+
+  public async run(): Promise<ServerVersionReponseDto> {
+    console.log('Getting server information');
+
+    await this.connect();
+    const { data: versionInfo } = await this.immichApi.serverInfoApi.getServerVersion();
+
+    console.log(versionInfo);
+    return versionInfo;
```
Comment (jrasm91) [inline cli/src/commands/server-info.ts:15]: Why do we need to log and return this information? Should it not be one of the other?

File: cli/src/commands/upload.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:1]: Is this line necessary?

File: cli/src/commands/upload.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+    if (crawledFiles.length === 0) {
+      console.log('No assets found, exiting');
+      return;
+    }
+
+    const assetsToUpload = crawledFiles.map((path) => new CrawledAsset(path));
+
+    const cliProgress = require('cli-progress');
+    const byteSize = require('byte-size');
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:42]: Why are these required instead of imported using import statements above?

File: cli/src/commands/upload.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        totalSizeUploaded += asset.fileSize;
+        uploadCounter++;
+      }
+
+      uploadProgress.update(sizeSoFar, { value_formatted: byteSize(sizeSoFar) });
+    }
+
+    uploadProgress.stop();
+
+    if (options.import) {
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:93]: I think we can simple this very nested if/else set of conditions. At least for the dry run stuff it looks like the messages are basically identical. Maybe we can get a message prefix of `Would have` vs `Successfully` and prepend that the the messages instead of using an if/else in some places.

File: cli/src/commands/upload.ts [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+      }
+    }
+  }
+
+  private async uploadAsset(asset: CrawledAsset, skipHash = false) {
+    await asset.readData();
+
+    let skipUpload = false;
+    if (!skipHash) {
+      const checksum: string = await asset.hash();
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:135]: `: string` should no be needed, it should be implied by the `.hash()` return type.

File: cli/src/commands/upload.ts [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+      skipUpload = checkResponse.data.results[0].action === 'reject';
+    }
+
+    if (skipUpload) {
+      asset.skipped = true;
+    } else {
+      const uploadFormData = new FormData();
+
+      uploadFormData.append('assetType', asset.assetType);
+      uploadFormData.append('assetData', asset.assetData, { filename: asset.path });
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:147]: Might be good to move this to end end of the form data.

File: cli/src/commands/upload.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+      }
+
+      if (!this.dryRun) {
+        await this.uploadService.upload(uploadFormData);
+      }
+    }
+  }
+
+  private async importAsset(asset: CrawledAsset) {
+    const importData = {
```
Comment (jrasm91) [inline cli/src/commands/upload.ts:169]: You can just inline this on line 180.

File: cli/src/cores/models/crawled-asset.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  public sidecarPath?: string;
+  public fileSize!: number;
+  public skipped = false;
+
+  constructor(path: string) {
+    this.path = path;
+  }
+
+  async readData() {
+    this.assetData = await fs.promises.readFile(this.path);
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:27]: This reads the whole file into memory, which is not going to work for large files. Maybe `assetData` can be a type stream, or whatever is returned from `fs.createReadStream`.

File: cli/src/cores/models/crawled-asset.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    const sideCarPath = `${this.path}.xmp`;
+    try {
+      fs.accessSync(sideCarPath, fs.constants.R_OK);
+    } catch (error) {
+      // No sidecar file
+      hasSidecar = false;
+    }
+    if (hasSidecar) {
+      this.sidecarPath = `${this.path}.xmp`;
+      this.sidecarData = await fs.promises.readFile(this.sidecarPath);
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:57]: These two lines can be right after line 50 instead of inside an if statement here.

File: cli/src/cores/models/crawled-asset.ts (hunk 14 above)
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:27]: This reads the whole file into memory, which is not going to work for large files. Maybe `assetData` can be a type stream, or whatever is returned from `fs.createReadStream`.

File: cli/src/cores/models/crawled-asset.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+  async process() {
+    const stats = await fs.promises.stat(this.path);
+    this.deviceAssetId = `${basename(this.path)}-${stats.size}`.replace(/\s+/g, '');
+
+    // TODO: Determine file type from extension only
+    const mimeType = mime.lookup(this.path);
+    if (!mimeType) {
+      throw Error('Cannot determine mime type of asset: ' + this.path);
+    }
+    this.assetType = mimeType.split('/')[0].toUpperCase() as AssetTypeEnum;
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:39]: TODO: move this to the server :-)

File: cli/src/cores/models/crawled-asset.ts (hunk 15 above)
Comment (jrasm91) [reply cli/src/cores/models/crawled-asset.ts:57]: hasSidecar is redundant now, since it is basically equal to `!!this.sidecarData`

File: cli/src/cores/models/crawled-asset.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    this.fileCreatedAt = stats.ctime.toISOString();
+    this.fileModifiedAt = stats.mtime.toISOString();
+    this.fileExtension = path.extname(this.path);
+    this.fileSize = stats.size;
+
+    let hasSidecar = true;
+
+    // TODO: doesn't xmp replace the file extension? Will need investigation
+    const sideCarPath = `${this.path}.xmp`;
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:48]: `sidecarPath`

File: cli/src/cores/models/crawled-asset.ts [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  async delete(): Promise<void> {
+    return fs.promises.unlink(this.path);
+  }
+
+  public async hash(): Promise<string> {
+    const hash = crypto.createHash('sha1');
+
+    const sha1 = (filePath: string) =>
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts:68]: Line 66 should be inside this method definition and the whole method should be moved outside `hash()` so it doesn't get redefined over and over.

File: cli/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+  .action((paths, options) => {
+    options.excludePatterns = options.ignore;
+    new Upload().run(paths, options);
+  });
+
+program
+  .command('import')
+  .description('Import existing assets')
+  .usage('[options] [paths...]')
+  .addOption(new Option('-r, --recursive', 'Recursive').env('IMMICH_RECURSIVE').default(false))
```
Comment (jrasm91) [inline cli/src/index.ts:31]: There are a decent amount of shared options between these two. Any way to define them once and use them in both places? Or have a function that can be called from both places to add them in one place?

File: cli/src/services/crawl.service.spec.ts [hunk 20]
```diff
@@ -0,0 +1,235 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (jrasm91) [inline cli/src/services/crawl.service.spec.ts:2]: We should get rid of these I think.

File: cli/src/services/crawl.service.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+      return crawledFiles;
+    } else {
+      searchPattern = '{' + directories.join(',') + '}';
+    }
+
+    if (crawlOptions.recursive) {
+      searchPattern = searchPattern + '/**/';
+    }
+
+    searchPattern = searchPattern + '/*.{' + ACCEPTED_FILE_EXTENSIONS.join(',') + '}';
```
Comment (jrasm91) [inline cli/src/services/crawl.service.ts:35]: We may want to consider alternative methods to read and generate this pattern. Raw string concatenation usually isn't the best...

File: cli/src/services/session.service.spec.ts [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+import { LoginError } from '../cores/errors/login-error';
+
+const mockPingServer = jest.fn(() => Promise.resolve({ data: { res: 'pong' } }));
+const mockUserInfo = jest.fn(() => Promise.resolve({ data: { email: 'admin@example.com' } }));
+
+jest.mock('../api/open-api', () => {
+  return {
+    __esModule: true,
+    ...jest.requireActual('../api/open-api'),
+    UserApi: jest.fn().mockImplementation(() => {
```
Comment (jrasm91) [inline cli/src/services/session.service.spec.ts:14]: You can also use `mockReturnValue({ getMyUserInfo: mockUserInfo })` I believe.

File: cli/src/services/session.service.ts [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+    });
+
+    console.log(`Logged in as ${userInfo.email}`);
+
+    if (!fs.existsSync(this.configDir)) {
+      // Create config folder if it doesn't exist
+      fs.mkdirSync(this.configDir, { recursive: true });
+    }
+
+    fs.writeFileSync(this.authPath, yaml.stringify({ instanceUrl: instanceUrl, apiKey: apiKey }));
```
Comment (jrasm91) [inline cli/src/services/session.service.ts:59]: If the key and value are the same you can just do this `({ instanceUrl, apiKey })`

File: cli/test-assets/animals/dinosaur.png
Comment (jrasm91) [inline cli/test-assets/animals/dinosaur.png:1]: do we need these still?

File: cli/package.json (hunk 5 above)
Comment (etnoy) [reply cli/package.json:7]: Hm...I tried removing systeminformation (was that the one you meant?) and it didn't compile. Don't know what it does, t though

File: cli/src/commands/server-info.ts (hunk 7 above)
Comment (etnoy) [reply cli/src/commands/server-info.ts:15]: Done

File: cli/src/services/session.service.ts (hunk 23 above)
Comment (etnoy) [reply cli/src/services/session.service.ts:59]: Done

File: cli/test-assets/animals/dinosaur.png
Comment (etnoy) [reply cli/test-assets/animals/dinosaur.png:1]: I haven't used them, so let's remove them

File: cli/README.md (hunk 4 above)
Comment (etnoy) [reply cli/README.md:1]: Yes, they will as soon as we feel this one is read

File: cli/package.json (hunk 5 above)
Comment (jrasm91) [reply cli/package.json:7]: Looks like we get some sort of identifier with it, maybe it's fine for now.

File: cli/package.json (hunk 6 above)
Comment (jrasm91) [reply cli/package.json]: I think we still need the build script, but it can just be `tsc -b` probably.

File: cli/src/commands/upload.ts (hunk 9 above)
Comment (etnoy) [reply cli/src/commands/upload.ts]: Done

File: cli/src/commands/upload.ts (hunk 8 above)
Comment (etnoy) [reply cli/src/commands/upload.ts]: Done

File: cli/src/commands/upload.ts (hunk 10 above)
Comment (etnoy) [reply cli/src/commands/upload.ts:91]: Yes, this turned out a bit messy in the end

File: cli/src/commands/upload.ts (hunk 11 above)
Comment (etnoy) [reply cli/src/commands/upload.ts:133]: Done

File: cli/src/commands/upload.ts (hunk 12 above)
Comment (etnoy) [reply cli/src/commands/upload.ts:145]: OK

File: cli/src/services/crawl.service.spec.ts (hunk 20 above)
Comment (jrasm91) [reply cli/src/services/crawl.service.spec.ts:2]: You should not be using `require()` to import a module.. Can you not get it to work with `import * as matchers from 'jest-extend'`?

File: cli/src/commands/upload.ts (hunk 13 above)
Comment (etnoy) [reply cli/src/commands/upload.ts:167]: Done

File: cli/src/cores/models/crawled-asset.ts (hunk 15 above)
Comment (etnoy) [reply cli/src/cores/models/crawled-asset.ts:57]: Done

File: cli/src/cores/models/crawled-asset.ts (hunk 14 above)
Comment (etnoy) [reply cli/src/cores/models/crawled-asset.ts]: Done

File: cli/src/cores/models/crawled-asset.ts (hunk 17 above)
Comment (etnoy) [reply cli/src/cores/models/crawled-asset.ts:48]: Done

File: cli/src/cores/models/crawled-asset.ts (hunk 14 above)
Comment (etnoy) [reply cli/src/cores/models/crawled-asset.ts]: Done

File: cli/src/services/crawl.service.spec.ts (hunk 20 above)
Comment (etnoy) [reply cli/src/services/crawl.service.spec.ts:2]: jest-extend seems to be very special. Even with that import line I get 

    Cannot find module 'jest-extend' or its corresponding type declarations.ts(2307)

File: cli/src/services/crawl.service.ts (hunk 21 above)
Comment (etnoy) [reply cli/src/services/crawl.service.ts:35]: It's a little bit neater now, but only a little

File: cli/src/index.ts (hunk 19 above)
Comment (etnoy) [reply cli/src/index.ts:31]: There's no easy way to do it, I decided that the cleanest way is the current one

File: cli/src/cores/models/crawled-asset.ts [hunk 24]
```diff
@@ -42,37 +42,30 @@ export class CrawledAsset {
...
-        rs.on('data', (chunk) => hash.update(chunk));
-        rs.on('end', () => resolve(hash.digest('hex')));
-      });
+  sha1 = (filePath: string) =>
+    new Promise<string>((resolve, reject) => {
+      const rs = fs.createReadStream(filePath);
+      rs.on('error', reject);
+      rs.on('data', (chunk) => this.hashFunction.update(chunk));
+      rs.on('end', () => resolve(this.hashFunction.digest('hex')));
+    });
```
Comment (jrasm91) [inline cli/src/cores/models/crawled-asset.ts]: Moving them here makes them class methods, which is OK... maybe something like this is better:

```typescript
const sha1 = (filePath: string) => {
    const hash = crypto.createHash('sha1');
    return new Promise<string>((resolve, reject) => {
      const rs = fs.createReadStream(filePath);
      rs.on('error', reject);
      rs.on('data', (chunk) => hash.update(chunk));
      rs.on('end', () => resolve(hash.digest('hex')));
    });
}
```
Or, if you want the method to be on the class still

```typescript
  private sha1(filePath: string) {
    const hash = crypto.createHash('sha1');
    return new Promise<string>((resolve, reject) => {
      const rs = fs.createReadStream(filePath);
      rs.on('error', reject);
      rs.on('data', (chunk) => hash.update(chunk));
      rs.on('end', () => resolve(hash.digest('hex')));
    });
  }
```

File: cli/src/services/session.service.spec.ts (hunk 22 above)
Comment (etnoy) [reply cli/src/services/session.service.spec.ts:14]: No, that doesn't work since the init order reverses

File: cli/src/cores/models/crawled-asset.ts (hunk 24 above)
Comment (etnoy) [reply cli/src/cores/models/crawled-asset.ts]: OK, did as per your suggestion

**Comment by neozhu** [comment]
Hi @etnoy,
 I'd like to propose a new feature. Could you add a 'trace' or 'monitor' cli command to monitor the upload directory? In this way, as soon as new photos are added to the directory, they could be automatically added to the album. One of my current use cases is that the photographer would turn on the WiFi function of the camera and upload the newly taken photos to the server via FTP in real-time. In this way, I can immediately add the photos to the album to achieve a live broadcast-like function. This feature is fantastic, isn't it? Of course, I know we could also achieve a similar function by periodically polling to execute the import command. However, I think adding a 'trace' or 'monitor' function would be better.

**Comment by alextran1502** [comment]
@etnoy , @jrasm91 If nothing else needs to be addressed in this PR, I can merge it in, and then we can start building from this foundation.

**Comment by etnoy** [comment]
I say let's merge and we'll take it from there.

Since I never used albums in immich I've not implemented that in the cli yet

**Comment by etnoy** [comment]
> Hi @etnoy, I'd like to propose a new feature. Could you add a 'trace' or 'monitor' cli command to monitor the upload directory? In this way, as soon as new photos are added to the directory, they could be automatically added to the album. One of my current use cases is that the photographer would turn on the WiFi function of the camera and upload the newly taken photos to the server via FTP in real-time. In this way, I can immediately add the photos to the album to achieve a live broadcast-like function. This feature is fantastic, isn't it? Of course, I know we could also achieve a similar function by periodically polling to execute the import command. However, I think adding a 'trace' or 'monitor' function would be better.

That will be done when the library pr is merged
